src/main/java/com/angelbroking/smartapi/SmartConnect.java

```diff
@@ -1,727 +1,781 @@
+     * @throws JSONException     if there is an error while parsing the JSON response
+     */
+    public User generateSession(String clientCode, String password, String totp) {
+        User user;
+        JSONObject loginResultObject;
+        try {
+            smartAPIRequestHandler = new SmartAPIRequestHandler(proxy);
+
+            loginResultObject = smartAPIRequestHandler.postRequest(this.apiKey, routes.getLoginUrl(), createLoginParams(clientCode, password, totp));
+
+        } catch (SmartAPIException ex) {
+            logger.error(Constants.SMART_API_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        } catch (IOException ex) {
+            logger.error(Constants.IO_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        } catch (JSONException ex) {
+            logger.error(Constants.JSON_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        }
+        if (loginResultObject.optJSONObject(Constants.SMART_CONNECT_DATA) == null) {
+            return null;
+        }
+        String jwtToken = loginResultObject.optJSONObject(Constants.SMART_CONNECT_DATA).optString(Constants.SMART_CONNECT_JWT_TOKEN);
+        String refreshTokenLocal = loginResultObject.optJSONObject(Constants.SMART_CONNECT_DATA).optString(Constants.SMART_CONNECT_REFRESH_TOKEN);
+        String feedToken = loginResultObject.optJSONObject(Constants.SMART_CONNECT_DATA).optString(Constants.SMART_CONNECT_FEED_TOKEN);
+        String url = routes.get(Constants.SMART_CONNECT_API_USER_PROFILE);
+        try {
+            user = ResponseParser.parseResponse(smartAPIRequestHandler.getRequest(this.apiKey, url, jwtToken));
+        } catch (SmartAPIException ex) {
+            logger.error(Constants.SMART_API_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        } catch (IOException ex) {
+            logger.error(Constants.IO_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        } catch (JSONException ex) {
+            logger.error(Constants.JSON_EXCEPTION_OCCURRED + ex.getMessage(), ex);
+            return null;
+        }
+        user.setAccessToken(jwtToken);
+        user.setRefreshToken(refreshTokenLocal);
+        user.setFeedToken(feedToken);
+
+        return user;
+
+    }
+
+    private JSONObject createLoginParams(String clientCode, String password, String totp) {
+        // Create JSON params object needed to be sent to api.
+
+        JSONObject params = new JSONObject();
+        params.put(Constants.SMART_CONNECT_CLIENT_CODE, clientCode);
+        params.put(Constants.SMART_CONNECT_PASSWORD, password);
+        params.put(Constants.SMART_CONNECT_TOTP, totp);
+        return params;
+    }
+
+    /**
+     * Hex encodes sha256 output for android support.
+     *
+     * @param str is the String that has to be encrypted.
+     * @return Hex encoded String.
+     */
+    public String sha256Hex(String str) {

```

Comment:
Can be moved to utis. Please review other methods as well.
Also please look into why have been these marked as public?

Reply:
have moved sha256Hex to utils class